File: comp-sci/maze-dfs/c/maze.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/time.h>

#define WIDTH   6
#define HEIGHT  6

struct node {
    int north_wall;
    int west_wall;
    int visited;
    int x;
    int y;
};

void printMaze(int cnt, struct node *cells)
{
    int y,x;
    for (y=0; y<HEIGHT; ++y) {
        for (x=0; x<WIDTH; ++x) {
            struct node *c = cells + ( x + y*WIDTH );
            printf("+");
            if (c->north_wall) {
                if (c->x == 0 && c->y == 0) 
                    printf("  ");
                else
                    printf("--");
            }
            else 
                printf("  ");
        }
        printf("+\n");

        for (x=0; x<WIDTH; ++x) {
            struct node *c = cells + (x + y*WIDTH) ;
            if (c->west_wall) 
                printf("|  ");
            else
                printf("   ");
        }
        printf("|\n");
    }
    
    for (x=0; x<WIDTH-1; ++x) {
        printf("+--");
    }
    printf("+  +\n\n");
}

// rtn is a pointer array.  NWES.
int neighbor(struct node *cells, struct node *cur, struct node **rtn)
{
    int x = cur->x;
    int y = cur->y;
    rtn[0] = rtn[1] = rtn[2] = rtn[3] = NULL;

    if (y-1 >= 0) rtn[0] = cells + (x + (y-1)*WIDTH);
    if (x-1 >= 0) rtn[1] = cells + ((x-1) + y*WIDTH);
    if (x+1 < WIDTH)  rtn[2] = cells + ((x+1) + y*WIDTH);
    if (y+1 < HEIGHT) rtn[3] = cells + (x + (y+1)*WIDTH);

    return rtn[0] != NULL || rtn[1] != NULL ||
           rtn[2] != NULL || rtn[3] != NULL;
}

void shuffle(struct node **array, size_t n) {    
    struct timeval tv;
    gettimeofday(&tv, NULL);
    int usec = tv.tv_usec;
    srand48(usec);

    if (n > 1) {
        size_t i;
        for (i = n - 1; i > 0; i--) {
            size_t j = (unsigned int) (drand48()*(i+1));
            struct node  *t = array[j];
            array[j] = array[i];
            array[i] = t;
        }
    }
}

void punchWall(struct node *node1, struct node *node2)
{
    int x1 = node1->x;
    int y1 = node1->y;
    int x2 = node2->x;
    int y2 = node2->y;

    if (x1 < x2 && y1 == y2)
        node2->west_wall =0;
    if (x2 < x1 && y1 == y2)
        node1->west_wall =0;

    if (y1 > y2 && x1 == x2)
        node1->north_wall =0;
    if (y2 > y1 && x1 == x2)
        node2->north_wall =0;
        
}

void genMaze(struct node *cells, struct node *cur)
{
    int i;
    if (cur->visited == 0) {
        struct node *nei[4];

        cur->visited = 1;
        if (neighbor( cells, cur, nei)) {
            shuffle(nei, 4);
            for (i=0; i<4; ++i) {
                if (nei[i] && nei[i]->visited==0) {
                    punchWall(cur, nei[i]);
                    genMaze(cells, nei[i]);
                }
            }
        }
    }
}

int main(int argc, char **argv) {
    int i;
    struct node cells[ WIDTH * HEIGHT ];

    // Initialize Cells
    for (i=0; i < WIDTH*HEIGHT; ++i) {
        cells[i].north_wall = cells[i].west_wall = 1;
        cells[i].visited = 0;
        cells[i].x = i % WIDTH;
        cells[i].y = i / WIDTH;
    }
    genMaze(cells, &cells[0]);
    printMaze(WIDTH * HEIGHT, cells);
    return 0;
}

File: comp-sci/maze-dfs/py/maze.py
import random

class MazeCell:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.north_wall = True
        self.west_wall = True
        self.visited = False

class Maze:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.cells = [ MazeCell(x, y) for y in range(height) for x in range(width) ]

    def get_cell(self, x, y):
        if 0 <= x < self.width and 0 <= y < self.height:
            return self.cells[x + y * self.width]
        return None

    def get_neighbors(self, cell):
        # Order:
        #  - 0: North,
        #  - 1: West,
        #  - 2: East,
        #  - 3: South
        neighbors = []
        
        coords = [(0, -1), (-1, 0), (1, 0), (0, 1)]
        for dx, dy in coords:
            neighbor = self.get_cell(cell.x + dx, cell.y + dy)
            neighbors.append(neighbor)
        return neighbors

    def punch_wall(self, node1, node2):
        x1, y1 = node1.x, node1.y
        x2, y2 = node2.x, node2.y

        if x1 < x2 and y1 == y2:
            node2.west_wall = False
        elif x2 < x1 and y1 == y2:
            node1.west_wall = False
        elif y1 > y2 and x1 == x2:
            node1.north_wall = False
        elif y2 > y1 and x1 == x2:
            node2.north_wall = False

    def generate(self, current_cell=None):
        if current_cell is None:
            current_cell = self.cells[0]
        
        current_cell.visited = True
        neighbors = self.get_neighbors(current_cell)
        
        # Filter out None and already visited neighbors
        valid_neighbors = [ n for n in neighbors if n is not None and \
                           not n.visited ]
        random.shuffle(valid_neighbors)
        for next_cell in valid_neighbors:
            if not next_cell.visited:
                self.punch_wall(current_cell, next_cell)
                self.generate(next_cell)


    def solve_trace(self, start_coords=(0, 0), end_coords=None):
        if end_coords is None:
            end_coords = (self.width - 1, self.height - 1)
        
        start_cell = self.get_cell(*start_coords)
        end_cell = self.get_cell(*end_coords)
        
        path = []
        visited = set()

        def dfs(curr):
            visited.add(curr)
            path.append(curr)
            # Yield the current path and current cell for animation
            yield list(path), curr

            if curr == end_cell:
                return True

            neighbors = self.get_neighbors(curr)
            # N, W, E, S
            checks = [
                (neighbors[0], curr.north_wall),  # North
                (neighbors[1], curr.west_wall),   # West
                (neighbors[2], neighbors[2].west_wall if neighbors[2] else True), # East
                (neighbors[3], neighbors[3].north_wall if neighbors[3] else True)  # South
            ]

            for neighbor, has_wall in checks:
                if neighbor and neighbor not in visited and not has_wall:
                    # Recursive yield from
                    found = yield from dfs(neighbor)
                    if found:
                        return True
                    # Backtracking: yield the state after returning from a dead end
                    yield list(path), curr

            path.pop()
            return False

        yield from dfs(start_cell)

    def solve(self, start_coords=(0, 0), end_coords=None):
        """Solve with DFS
        NOTE: this solve() can only solve Perfect Maze:
        Where Perfect maze:
          - All nodes are connected to the root
          - Loop is not made.
          - There is only 1 (unique) path from root to the goal node.
            - can be represented as Tree
        """
        if end_coords is None:
            end_coords = (self.width - 1, self.height - 1)
        
        start_cell = self.get_cell(*start_coords)
        end_cell = self.get_cell(*end_coords)
        
        path = []
        visited = set()

        def dfs(curr):
            """depth first search for Maze
            nested function to be convenient, and scoped.
            """
            if curr == end_cell:
                path.append(curr)
                return True
            
            visited.add(curr)
            path.append(curr)

            # Check neighbors and connectivity
            # Order:
            #  - 0: North,
            #  - 1: West,
            #  - 2: East,
            #  - 3: South

            neighbors = self.get_neighbors(curr)
            north, west, east, south = 0,1,2,3
            
            n = neighbors[north]
            if n and n not in visited and not curr.north_wall:
                if dfs(n): return True
            w = neighbors[west]
            if w and w not in visited and not curr.west_wall:
                if dfs(w): return True
            e = neighbors[east]
            if e and e not in visited and not e.west_wall:
                if dfs(e): return True
            s = neighbors[south]
            if s and s not in visited and not s.north_wall:
                if dfs(s): return True

            path.pop()
            return False

        dfs(start_cell)
        return path

    def __str__(self):
        rtn = []
        for y in range(self.height):
            north_row = ""
            for x in range(self.width):
                cell = self.get_cell(x, y)
                north_row += "+"
                if cell.north_wall:
                    if cell.x == 0 and cell.y == 0:
                        north_row += "  "
                    else:
                        north_row += "--"
                else:
                    north_row += "  "
            rtn.append(north_row + "+")
            
            west_row = ""
            for x in range(self.width):
                cell = self.get_cell(x, y)
                if cell.west_wall:
                    west_row += "|  "
                else:
                    west_row += "   "
            rtn.append(west_row + "|")

        bottom = ""
        for x in range(self.width - 1):
            bottom += "+--"
        rtn.append(bottom + "+  +")

        return "\n".join(rtn)
